### ISSUE TITLE
OpenSSL x86-windows-static not working with find_package(OpenSSL REQUIRED)

### ISSUE THREAD

**Issue by 67-al-ahad** [microsoft/vcpkg#12001]
**Describe the bug**
A basic openssl project is not able to be generated and compiled due to the lack of vcpkg to find the library paths.

**Environment**
- OS: Win 10
- Compiler: MSVC++ 2019 

**To Reproduce**
1. ./vcpkg install openssl:x86-windows-static-md
2. OPENSSL_ROOT_DIR is not detected automatically

**Expected behavior**
The generation should be without any problem

**Failure logs** 
```
Could NOT find OpenSSL, try to set the path to OpenSSL root folder in the
 system variable OPENSSL_ROOT_DIR (missing: OPENSSL_LIBRARIES)
```

**Additional context**
It is working well when we are installing **openssl:x86-windows** version of the library. But as soon as we don't want to use the DLL version instead **openssl:x86-windows-static-md** then nothing works anymore.

[helloworld.zip](https://github.com/microsoft/vcpkg/files/4796001/helloworld.zip)

**Comment by JackBoosY** [comment]
See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/users/integration.md#with-cmake).

**Comment by 67-al-ahad** [comment]
Ok although I set the correct triplets in CMakeSettings.json

```
{
  "configurations": [
    {
      "name": "x86-Debug",
      "generator": "Ninja",
      "configurationType": "Debug",
      "buildRoot": "${projectDir}\\out\\build\\${name}",
      "installRoot": "${projectDir}\\out\\install\\${name}",
      "cmakeCommandArgs": "-DVCPKG_TARGET_TRIPLET=x86-windows-static",
      "buildCommandArgs": "",
      "ctestCommandArgs": "",
      "inheritEnvironments": [ "msvc_x86" ],
      "variables": [],
      "cmakeToolchain": "C:\\vcpkg\\scripts\\buildsystems\\vcpkg.cmake"
    },
    {
      "name": "x86-Release",
      "generator": "Ninja",
      "configurationType": "RelWithDebInfo",
      "buildRoot": "${projectDir}\\out\\build\\${name}",
      "installRoot": "${projectDir}\\out\\install\\${name}",
      "cmakeCommandArgs": "-DVCPKG_TARGET_TRIPLET=x86-windows-static",
      "buildCommandArgs": "",
      "ctestCommandArgs": "",
      "inheritEnvironments": [ "msvc_x86" ],
      "variables": [],
      "cmakeToolchain": "C:\\vcpkg\\scripts\\buildsystems\\vcpkg.cmake"
    }
  ]
}
```

I still get this error

```
[1/1] Linking CXX executable MonaTiny\MonaTiny.exe
  FAILED: MonaTiny/MonaTiny.exe 
  cmd.exe /C "cd . && "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_exe --intdir=MonaTiny\CMakeFiles\MonaTiny.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x86\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x86\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1426~1.288\bin\Hostx86\x86\link.exe /nologo MonaTiny\CMakeFiles\MonaTiny.dir\sources\MonaTiny.cpp.obj MonaTiny\CMakeFiles\MonaTiny.dir\sources\main.cpp.obj  /out:MonaTiny\MonaTiny.exe /implib:MonaTiny\MonaTiny.lib /pdb:MonaTiny\MonaTiny.pdb /version:0.0  /machine:X86 /debug /INCREMENTAL /subsystem:console  MonaBase\MonaBased.lib  MonaCore\MonaCored.lib  C:\vcpkg\installed\x86-windows-static\debug\lib\libssl.lib  C:\vcpkg\installed\x86-windows-static\debug\lib\libcrypto.lib  MonaBase\MonaBased.lib  C:\vcpkg\installed\x86-windows-static\debug\lib\libssl.lib  C:\vcpkg\installed\x86-windows-static\debug\lib\libcrypto.lib  C:\vcpkg\installed\x86-windows-static\debug\lib\libssl.lib  C:\vcpkg\installed\x86-windows-static\debug\lib\libcrypto.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cmd.exe /C "cd /D G:\programming\c++\Mona\67-al-ahad\out\build\x86-Debug\MonaTiny && powershell -noprofile -executionpolicy Bypass -file C:/vcpkg/scripts/buildsystems/msbuild/applocal.ps1 -targetBinary G:/programming/c++/Mona/67-al-ahad/out/build/x86-Debug/MonaTiny/MonaTiny.exe -installedDir C:/vcpkg/installed/x86-windows-static/debug/bin -OutVariable out""
  LINK Pass 1: command "C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1426~1.288\bin\Hostx86\x86\link.exe /nologo MonaTiny\CMakeFiles\MonaTiny.dir\sources\MonaTiny.cpp.obj MonaTiny\CMakeFiles\MonaTiny.dir\sources\main.cpp.obj /out:MonaTiny\MonaTiny.exe /implib:MonaTiny\MonaTiny.lib /pdb:MonaTiny\MonaTiny.pdb /version:0.0 /machine:X86 /debug /INCREMENTAL /subsystem:console MonaBase\MonaBased.lib MonaCore\MonaCored.lib C:\vcpkg\installed\x86-windows-static\debug\lib\libssl.lib C:\vcpkg\installed\x86-windows-static\debug\lib\libcrypto.lib MonaBase\MonaBased.lib C:\vcpkg\installed\x86-windows-static\debug\lib\libssl.lib C:\vcpkg\installed\x86-windows-static\debug\lib\libcrypto.lib C:\vcpkg\installed\x86-windows-static\debug\lib\libssl.lib C:\vcpkg\installed\x86-windows-static\debug\lib\libcrypto.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:MonaTiny\CMakeFiles\MonaTiny.dir/intermediate.manifest MonaTiny\CMakeFiles\MonaTiny.dir/manifest.res" failed (exit code 1104) with the following output:
G:\programming\c++\Mona\67-al-ahad\out\build\x86-Debug\LINK : fatal error LNK1104: cannot open file 'libcryptod.lib'
  ninja: build stopped: subcommand failed.
```

![image](https://user-images.githubusercontent.com/66976729/85025689-feaa0000-b177-11ea-96d1-ceef88aaf8e0.png)

The error is here the file name installed by vcpkg is **C:\vcpkg\installed\x86-windows-static\debug\lib\libcrypto.lib ** and not **libcryptod.lib **

**Comment by 67-al-ahad** [comment]
What is this error linked to? My code or the OpenSSL port?

**Comment by 67-al-ahad** [comment]
Any idea what is the problem here ?

**Comment by JackBoosY** [comment]
@67-al-ahad Can you confirm your vcpkg is latest?
Openssl should generated `libeay32.lib` and `ssleay32.lib`.
And the triplet is community rather than official, so we do not provide maintenance for it.

Thanks.

**Comment by 67-al-ahad** [comment]
Vcpkg package management program version 2020.02.04-nohash

See LICENSE.txt for license information.

**Comment by 67-al-ahad** [comment]
When I copy the 2 files and add the **d** letter at the end and click the **Save and generate CMake cache to load variables**, Visual Studio seems to be able to now recognize the correct files.

![image](https://user-images.githubusercontent.com/66976729/85400545-deec5080-b558-11ea-9f6d-e4eed490fe7f.png)

**Comment by 67-al-ahad** [comment]
I added the explanation for this bug and the workaround here:

https://github.com/67-al-ahad/MonaServer2/blob/master/README.md

**Comment by Septarius** [comment]
The old OpenSSL packages for < 1.1.0 (ssleay32 and libeay32) might have had a "d" suffix or at least I could find hints at that. It appears that they removed that for 1.1+ when the name was changed to libcrypto and libssl. I would suggestion removing the _DEBUG pragma in your header file. That should clear your problem up. 

 **Release**
```
Directory: C:\Users\packages\vcpkg\installed\x86-windows\lib

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----         6/25/2020   8:27 PM        1014802 libcrypto.lib
-a----         6/25/2020   8:27 PM         122290 libssl.lib

 Directory: C:\Users\packages\vcpkg\installed\x86-windows-static\lib


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----         6/26/2020  12:44 AM       17505988 libcrypto.lib
-a----         6/26/2020  12:44 AM        3570946 libssl.lib
-a----         6/26/2020  12:44 AM        1069056 ossl_static.pdb
```

**Debug**
```
Directory: C:\Users\packages\vcpkg\installed\x86-windows\debug\lib

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----         6/25/2020   8:28 PM        1014802 libcrypto.lib
-a----         6/25/2020   8:28 PM         122290 libssl.lib

Directory: C:\Users\packages\vcpkg\installed\x86-windows-static\debug\lib

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----         6/26/2020  12:45 AM       14386074 libcrypto.lib
-a----         6/26/2020  12:45 AM        2916922 libssl.lib
-a----         6/26/2020  12:45 AM        1069056 ossl_static.pdb
```

**Comment by 67-al-ahad** [comment]
I am using vcpkg, CMake and Visual Studio 2019 are you talking about this : https://stackoverflow.com/questions/4604283/automatic-defines-according-to-debug-release-config-in-visual-studio

**Comment by Septarius** [comment]
Yeah the last one
```
#if defined(_DEBUG)
#pragma comment(lib, "libcryptod.lib")
#pragma comment(lib, "libssld.lib")
```
I noticed that there was a reference to libcryptod.lib in the header file but it wasn't cooperating at the time to quote it.

**Comment by JackBoosY** [comment]
@cenit Could you please take a look?

Thanks.

**Comment by 67-al-ahad** [comment]
> Yeah the last one
> 
> ```
> #if defined(_DEBUG)
> #pragma comment(lib, "libcryptod.lib")
> #pragma comment(lib, "libssld.lib")
> ```
> 
> I noticed that there was a reference to libcryptod.lib in the header file but it wasn't cooperating at the time to quote it.

Even after removing the **d** from header I still get this error.

**Comment by cenit** [comment]
@jackboosy @67-al-ahad there is a problem with the openssl port on vcpkg, which is independent from 1.1 update. The only culprit of that PR is not having fixed this problem and not even highlighted that.
On windows, as can be seen here 
https://github.com/Kitware/CMake/blob/45542049563f0ee8358ead563a4e4cc7d7270110/Modules/FindOpenSSL.cmake#L209-L221
`LIB_EAY_DEBUG` should have a trailing `d`. The same is valid for `SSL_EAY_DEBUG`. Unfortunately, our port does not deal with it. A wrapper might be required, in order to help cmake module finding it properly

**Comment by christophe-calmejane** [comment]
I'm facing a more critical issue regarding the OpenSSL port:
When generating a cmake solution, only the `debug` path is found and configured, resulting in the debug version of openSSL being linked then the openSSL DLL being used, which require the `debug vc runtime` which we are not allowed to distribute.

See the cmake log:
```
-- Found OpenSSL: D:/dev/vcpkg/installed/x86-windows/debug/lib/libcrypto.lib (found suitable version "1.1.1k", minimum required is "1")
```

ZLIB for example, correctly sets `debug` and `optimized` paths, so cmake knows which one to take based on the current config:
```
-- Found ZLIB: optimized;D:/dev/vcpkg/installed/x86-windows/lib/zlib.lib;debug;D:/dev/vcpkg/installed/x86-windows/debug/lib/zlibd.lib (found suitable version "1.2.11", minimum required is "1")
```

Not sure if https://github.com/microsoft/vcpkg/pull/21641 is trying to fix this specific issue.

**Comment by dg0yt** [comment]
@christophe-calmejane #21641 is an extension of a test port that I created to verify vcpkg's interaction with CMake's find modules.
An earlier revision of that PR confirmed that something is wrong with OpenSSL when consumed in CMake 3.4.3 (but not in CMake 3.21): https://dev.azure.com/vcpkg/public/_build/results?buildId=63856&view=logs&j=3859ca13-59fe-57bd-13f9-449df7d191a6&t=03b21fcd-db83-5643-e95b-35d3afaab803

I decided to move this to a separate PR (WIP) where I will also update port openssl to hopefully fix the issues. 

Related: #21415

Disclaimer: I test MSVC only in vcpkg CI.
